torch/csrc/jit/codegen/cuda/lower_index.cpp

```diff
@@ -172,89 +185,113 @@ void IndexLowering::handle(const ReductionOp* rop) {
   const auto out_tv = rop->out()->as<TensorView>();
   const auto out_domain = out_tv->domain();
 
-  const bool is_block_reduce = out_domain->hasBlockReduction();
-  const bool is_grid_reduce = out_domain->hasGridReduction();
-
-  // If we do a grid reduction we can't have a reduction axis that is not bound
-  // to a grid or block dim ()
-  if (is_grid_reduce) {
-    TORCH_INTERNAL_ASSERT(
-        std::none_of(
-            out_domain->domain().begin(),
-            out_domain->domain().end(),
-            [](IterDomain* id) {
-              return !id->isThread() && id->isReduction() &&
-                  !id->extent()->isOneInt();
-            }),
-        "Found a reduction stage that has both a non-parallelized ",
-        "reduction and a grid reduction.  This is not supported, ",
-        "please use rfactor to do the serialized reduction first, ",
-        "then the grid reduction.");
-  }
+  const bool has_block_reduce = out_domain->hasBlockReduction();
+  const bool has_grid_reduce = out_domain->hasGridReduction();
 
   const auto out = lowerDstIndex(rop->out());
   const auto in = lowerSrcIndex(rop->in(), rop->out());
 
-  ReductionOp* block_reduction_op = nullptr;
+  // Serial reduction
+  if (!has_block_reduce && !has_grid_reduce) {
+    pushBack(
+        IrBuilder::create<BinaryOp>(rop->getReductionOpType(), out, out, in));
+    return;
+  }
 
-  if (is_block_reduce) {
-    block_reduction_op = IrBuilder::create<ReductionOp>(
-        rop->getReductionOpType(), rop->init(), out, in);
-    if (rop->predicate()) {
-      block_reduction_op->setPredicate(rop->predicate());
-    }
-    if (rop->writePredicate()) {
-      block_reduction_op->setWritePredicate(rop->writePredicate());
-    }
-    pushBack(block_reduction_op);
+  ReductionOp* new_rop = IrBuilder::create<ReductionOp>(
+      rop->getReductionOpType(), rop->init(), out, in, rop->isFused());
+  if (rop->predicate()) {
+    new_rop->setPredicate(rop->predicate());
+  }
+  if (rop->writePredicate()) {
+    new_rop->setWritePredicate(rop->writePredicate());
   }
 
-  if (is_grid_reduce) {
-    const auto reduce_buffer = allocGlobalBufferForGridComm(
-        getGridCommWorkBufferSize(out_domain), out->dtype(), false);
+  // If not grid reduction, just append the new ReductionOp node
+  if (!has_grid_reduce) {
+    pushBack(new_rop);
+    return;
+  }
 
-    const auto sync_buffer = allocGlobalBufferForGridComm(
-        getGridSyncBufferSize(out_domain), DataType::Int, true);
+  handleGridReduction(new_rop);
+}
 
-    const auto grid_reduction_op = (block_reduction_op == nullptr)
-        ? IrBuilder::create<ReductionOp>(
-              rop->getReductionOpType(), rop->init(), out, in)
-        : block_reduction_op;
+void IndexLowering::handleGridReduction(ReductionOp* new_rop) {
```

Comment:
It may clean things up around here by having `BlockReduction` and have `BlockReduction` and `GridReduction` to inherit from `ReductionOp`.